File: viz.py
import Tkinter
import framebuf

class Viz(object):
    def __init__(self, parent):
        self.height = 650
        self.width = 1430
        self.xoffset = 36
        self.yoffset = 36
        self.xscale = 36
        self.yscale = 36
        self.radius = 6
        self.canvas = Tkinter.Canvas(parent, height=self.height, width=self.width, bg='#303030')
        self.canvas.grid(row=0)

        # draw the T
        self.canvas.create_line(self.xform(-0.5, -0.5), self.xform(8.5, -0.5), self.xform(8.5, 2.5),  self.xform(5.5, 2.5), \
            self.xform(5.5, 16.5), self.xform(2.5, 16.5), self.xform(2.5, 2.5), self.xform(-0.5, 2.5), self.xform(-0.5, -0.5), \
            fill='white')

        # draw a B
        b = 9.5
        self.canvas.create_line(self.xform(b, -0.5), self.xform(b+7.5, -0.5), self.xform(b+9, 1), self.xform(b+9, 7.5), \
            self.xform(b+8.5, 8.0), self.xform(b+9, 8.5), self.xform(b+9, 15), self.xform(b+7.5, 16.5), self.xform(b, 16.5), \
            self.xform(b, -0.5),\
            fill='white')
        y0 = 2.5
        self.canvas.create_line(self.xform(b+3, y0), self.xform(b+6, y0), self.xform(b+6, y0+4), self.xform(b+3, y0+4), \
            self.xform(b+3, y0), fill='white')
        y0 = y0 + 7
        self.canvas.create_line(self.xform(b+3, y0), self.xform(b+6, y0), self.xform(b+6, y0+4), self.xform(b+3, y0+4), \
            self.xform(b+3, y0), fill='white')

        # draw the G
        self.canvas.create_line(self.xform(21, -0.5), self.xform(28.5, -0.5), self.xform(28.5, 2.5), self.xform(22.5, 2.5), \
            self.xform(22.5, 13.5), self.xform(25.5, 13.5), self.xform(25.5, 6.5), self.xform(28.5, 6.5), self.xform(28.5, 16.5), \
            self.xform(21, 16.5), self.xform(19.5, 15.0), self.xform(19.5, 1), self.xform(21, -0.5), fill='white')

        # draw another B
        b = 29.5
        self.canvas.create_line(self.xform(b, -0.5), self.xform(b+7.5, -0.5), self.xform(b+9, 1), self.xform(b+9, 7.5), \
            self.xform(b+8.5, 8.0), self.xform(b+9, 8.5), self.xform(b+9, 15), self.xform(b+7.5, 16.5), self.xform(b, 16.5), \
            self.xform(b, -0.5),\
            fill='white')
        y0 = 2.5
        self.canvas.create_line(self.xform(b+3, y0), self.xform(b+6, y0), self.xform(b+6, y0+4), self.xform(b+3, y0+4), \
            self.xform(b+3, y0), fill='white')
        y0 = y0 + 7
        self.canvas.create_line(self.xform(b+3, y0), self.xform(b+6, y0), self.xform(b+6, y0+4), self.xform(b+3, y0+4), \
            self.xform(b+3, y0), fill='white')

    def visible(self, x, y):
        """is pixel x, y visible?"""
        if x >= 0 and x <= 2 and y >= 3:
            return False
        if x >= 6 and x <= 8 and y >= 3:
            return False
        if x == 9:
            return False
        if x >= 13 and x <= 15 and y >= 3 and y <= 6:
            return False
        if x >= 13 and x <= 15 and y >= 10 and y <= 13:
            return False
        if (x == 18 or x == 20) and (y == 0 or y == 16):
            return False
        if (x == 18 and y == 8):
            return False
        if (x == 19):
            return False
        if (x >= 23 and x <= 25) and (y >= 3 and y <= 13):
            return False
        if (x >= 26 and x <= 28 and (y >= 3 and y <= 6)):
            return False
        if (x == 29):
            return False
        if x >= 33 and x <= 35 and y >= 3 and y <= 6:
            return False
        if x >= 33 and x <= 35 and y >= 10 and y <= 13:
            return False
        if x == 38 and (y == 0 or y == 8 or y == 16):
            return False
        return True

    def xform(self, x, y):
        """map pixel in LED space to canvas coordinates"""
        cx = self.xoffset + self.xscale * x
        cy = self.yoffset + self.yscale * y
        return cx, cy

    def rgb2color(self, red, green, blue):
        """convert (r, g, b) into a # string"""
        hexcode = "#%02x%02x%02x" % (red, green, blue)
        return hexcode

    def plotpix(self, x, y, red, green, blue):
        """plot the LED pixel at x,y (pixel space) with color r,g,b"""
        if not self.visible(x, y):
            return
        cx, cy = self.xform(x, y)
        r = self.radius
        self.canvas.create_oval(cx-r, cy-r, cx+r, cy+r, fill=self.rgb2color(red, green, blue))
        #print "plot x="+str(x)+" y="+str(y)+" r="+str(red)+" g="+str(green)+" b="+str(blue)

    def plottrip(self, x, y, triple):
        self.plotpix(x, y, triple[0], triple[1], triple[2])

    def render(self, fb):
        """render the framebuf"""
        for x in range(0, fb.width()):
            for y in range(0, fb.height()):
                self.plottrip(x, y, fb.get(x, y))

File: viz.cpp
#include <cairomm/context.h>
#include <iostream>
#include <chrono>
#include <array>
#include "global.h"
#include "viz.h"

namespace { // hi Jeff

// TODO figure all this out on the fly someday
static const int XOFFSET = 38;
static const int YOFFSET = 36;
static const int XSCALE = 33;
static const int YSCALE = 36;
static const int RECTANGLE = 14;

struct P
{
    double x, y;
};

// letter outlines
// T (absolute x and y)
const std::vector<P> T = { {-0.5, -0.5}, {8.5, -0.5}, {8.5, 2.5}, {5.5, 2.5},
                           {5.5, 16.5}, {2.5, 16.5}, {2.5, 2.5}, {-0.5, 2.5} };
// outer part of B (relative x, absolute y)
const std::vector<P> Bo = { {0, -0.5}, {7.5, -0.5}, {9, 1}, {9, 7.5},
                            {8.5, 8.0}, {9, 8.5}, {9, 15}, {7.5, 16.5}, {0, 16.5}, };
// inner part of B (relative x and y)
const std::vector<P> Bi = { {3, 0}, {6, 0}, {6, 4}, {3, 4} };
// G (absolute x and y)
const std::vector<P> G = { {21, -0.5}, {28.5, -0.5}, {28.5, 2.5}, {22.5, 2.5},
                           {22.5, 13.5}, {25.5, 13.5}, {25.5, 6.5}, {28.5, 6.5}, {28.5, 16.5},
                           {21, 16.5}, {19.5, 15.0}, {19.5, 1} };

}

Viz::Viz(Framebuf& fb) : m_fb(fb), m_animation_name("none yet")
{
    set_size_request(1335, 650); // TODO constants
}

Viz::~Viz()
{
}

void
Viz::render()
{
    queue_draw();
}

bool
Viz::on_draw(const Cairo::RefPtr<Cairo::Context>& cr)
{
    auto start = std::chrono::system_clock::now();

    Gtk::Allocation allocation = get_allocation();
    const int width = allocation.get_width();
    const int height = allocation.get_height();
    //std::cout << "on_draw " << width << "x" << height << std::endl;

    // set gray background
    cr->set_line_width(1);
    cr->set_source_rgb(0.25, 0.25, 0.25);
    cr->rectangle(0, 0, width, height);
    cr->fill();
    cr->stroke();

    // SOMEDAY there is a better way to do this without all the constants, etc
    cr->set_source_rgb(0.5, 0.5, 0.5); // gray
    // draw a T
    for (int i = 0; i < T.size(); ++i)
    {
        const P& p = T[i];
        if (i == 0)
            cr->move_to(xformx(p.x), xformy(p.y));
        else
            cr->line_to(xformx(p.x), xformy(p.y));
    }
    cr->line_to(xformx(T[0].x), xformy(T[0].y));
    cr->stroke();

    // draw a B
    double b = 9.5;
    for (int i = 0; i < Bo.size(); ++i)
    {
        const P& p = Bo[i];
        if (i == 0)
            cr->move_to(xformx(b+p.x), xformy(p.y));
        else
            cr->line_to(xformx(b+p.x), xformy(p.y));
    }
    cr->line_to(xformx(b+Bo[0].x), xformy(Bo[0].y));
    cr->stroke();
    double y0 = 2.5;
    for (int i = 0; i < Bi.size(); ++i)
    {
        const P& p = Bi[i];
        if (i == 0)
            cr->move_to(xformx(b+p.x), xformy(y0+p.y));
        else
            cr->line_to(xformx(b+p.x), xformy(y0+p.y));
    }
    cr->line_to(xformx(b+Bi[0].x), xformy(y0+Bi[0].y));
    cr->stroke();
    y0 = 9.5;
    for (int i = 0; i < Bi.size(); ++i)
    {
        const P& p = Bi[i];
        if (i == 0)
            cr->move_to(xformx(b+p.x), xformy(y0+p.y));
        else
            cr->line_to(xformx(b+p.x), xformy(y0+p.y));
    }
    cr->line_to(xformx(b+Bi[0].x), xformy(y0+Bi[0].y));
    cr->stroke();

    // draw a G
    for (int i = 0; i < G.size(); ++i)
    {
        const P& p = G[i];
        if (i == 0)
            cr->move_to(xformx(p.x), xformy(p.y));
        else
            cr->line_to(xformx(p.x), xformy(p.y));
    }
    cr->line_to(xformx(G[0].x), xformy(G[0].y));
    cr->stroke();

    // draw another B
    b = 29.5;
    for (int i = 0; i < Bo.size(); ++i)
    {
        const P& p = Bo[i];
        if (i == 0)
            cr->move_to(xformx(b+p.x), xformy(p.y));
        else
            cr->line_to(xformx(b+p.x), xformy(p.y));
    }
    cr->line_to(xformx(b+Bo[0].x), xformy(Bo[0].y));
    cr->stroke();
    y0 = 2.5;
    for (int i = 0; i < Bi.size(); ++i)
    {
        const P& p = Bi[i];
        if (i == 0)
            cr->move_to(xformx(b+p.x), xformy(y0+p.y));
        else
            cr->line_to(xformx(b+p.x), xformy(y0+p.y));
    }
    cr->line_to(xformx(b+Bi[0].x), xformy(y0+Bi[0].y));
    cr->stroke();
    y0 = 9.5;
    for (int i = 0; i < Bi.size(); ++i)
    {
        const P& p = Bi[i];
        if (i == 0)
            cr->move_to(xformx(b+p.x), xformy(y0+p.y));
        else
            cr->line_to(xformx(b+p.x), xformy(y0+p.y));
    }
    cr->line_to(xformx(b+Bi[0].x), xformy(y0+Bi[0].y));
    cr->stroke();

    // DRAW THE PIXELS!
    double power = 0;
    int pixels = 0;
    {
        std::lock_guard<std::mutex> lock(m_fb.mutex());
        for (int x = 0; x < TBGB_XMAX; x++)
        {
            for (int y = 0; y < TBGB_YMAX; y++)
            {
                if (!::is_visible(x, y)) // oops, I picked a name that was already in use by gtk
                {
                    continue;
                }
                //cr->arc(xformx(x), xformy(y), RADIUS, 0, 2 * M_PI); circle, but squares are faster
                cr->set_source_rgb(m_fb.data(x, y).get_red(), m_fb.data(x, y).get_green(), m_fb.data(x, y).get_blue());
                power += (m_fb.data(x, y).get_red() + m_fb.data(x, y).get_green() + m_fb.data(x, y).get_blue());
                pixels++;
                cr->rectangle(xformx(x)-RECTANGLE/2, xformy(y)-RECTANGLE/2, RECTANGLE/2, RECTANGLE/2);
                cr->fill_preserve();
                cr->stroke();
            }
        }
    }
    cr->move_to(10, xformy(17));
    cr->set_source_rgb(1, 1, 1);
    char buf[100];
    snprintf(buf, sizeof(buf), "power: %6.2f%% animation: %s", power / 3 / pixels * 100, m_animation_name.c_str());
    //std::cout << "[" << buf << "] power=" << power << " pixels=" << pixels << " total=" << power / 3 / pixels << std::endl;
    cr->select_font_face("Courier", Cairo::FONT_SLANT_NORMAL, Cairo::FONT_WEIGHT_NORMAL);
    cr->show_text(buf);
    cr->stroke();
#if 0
    // diagonal for debugging screen size/paint stuff
    cr->set_source_rgba(0.337, 0.612, 0.117, 0.9);   // green
    cr->move_to(0, 0);
    cr->line_to(width-1, height-1);
    cr->stroke();
#endif

#if 0
    // timing
    auto end = std::chrono::system_clock::now();
    std::chrono::duration<double> diff = end - start;
    std::cout << "elapsed on_draw time: " << diff.count() << std::endl;
#endif
    return true;
}

double
Viz::xformx(double x)
{
    return XOFFSET + XSCALE * x;
}

double
Viz::xformy(double y)
{
    return YOFFSET + YSCALE * y;
}
